### PR TITLE
Introduce a Long-track bonus in CkfTrackBuilder

### PR DESCRIPTION
**Pull request by VinInn**
I have introduced a bonus mechanism for candiate comparison in CkfTrackBuilder similar to the one present since long time in the Grouped one.

The main effect should be to have longer electrons.
In this PR the default areno-op so no regression is expected.
Muon and HLT has been left with no Bonus mechanism as well.
 An optimization procedures can be applied for Electrons, conversions and eventually Muons to determine the parameters for production .

The effect can be observed in the number of missing outer hits for electrons in zee events with pile-up
(here for two values of the bonus 5 and 10 (10 is the default in grouped)
http://innocent.home.cern.ch/innocent/RelVal/zeePu14Bonus/plots_gsf_electronGsf/hitsAndPt.pdf

Discussion scheduled in TRK-POG meeting on June 3rd.

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by VinInn** [comment]
The DQM regressions are expected and expert need to validate at some point (after retuning?). I do not see the difference in TriggerResults: are all 0.

**Comment by mmusich** [comment]
> the test in question passed in my local area

It's likely a (transient) and unrelated connection glitch to confDB (at the same time of the failure, I observed other related failures).

**Comment by mmusich** [comment]
> . I do not see the difference in TriggerResults: are all 0.

See https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-05-22-2300+d0b7d4/62856/triggerResults/141.044_RunJetMET2023D/reHLT.log

File: RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc [hunk 1]
```diff
@@ -48,6 +50,8 @@ void CkfTrajectoryBuilder::fillPSetDescription(edm::ParameterSetDescription& iDe
   BaseCkfTrajectoryBuilder::fillPSetDescription(iDesc);
   iDesc.add<int>("maxCand", 5);
   iDesc.add<double>("lostHitPenalty", 30.);
+  iDesc.add<double>("foundHitBonus", 10.);
+  iDesc.add<int>("minHitForDoubleBonus", 8);
```
Comment (slava77) [inline RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:54]: do I understand correctly that with `"foundHitBonus", 0.` there will be no differences?

Was the non-zero value here added just to run the tests?
In the email thread I was proposing to make a technical PR.

Comment (VinInn) [reply RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:54]: 0 will be NoOp yes.
I can change default to NoOp if requested.

Comment (slava77) [reply RecoTracker/CkfPattern/src/CkfTrajectoryBuilder.cc:54]: > 0 will be NoOp yes. I can change default to NoOp if requested.

I may be more comfortable with a physics change after seeing some more complicated electron sources (including fakes). ttbar could be a good test.

**Comment by VinInn** [comment]
@cmsbuid , please test

**Comment by VinInn** [comment]
No regression is expected after fixing the defaults.

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 2]
```diff
@@ -268,6 +268,14 @@ def customizeHLTfor44576(process):
         break
     return process
 
+
+def customizeHLTfor45023(process):
+    for pset in process.psets.values():
+        if hasattr(pset, 'ComponentType') and pset.ComponentType == 'CkfTrajectoryBuilder':
+           pset.foundHitBonus = cms.double(0.0)
+           pset.minHitForDoubleBonus = cms.int32(9999)
+    return process
```
Comment (mmusich) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py:277]: is the customization still necessary after the last round of updates? i.e. IIUC the current defaults in `fillDescriptions` will ensure there is no regression at HLT?

Comment (VinInn) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:277]: In the old days tests would not pass as the HLT configs in the release do not contain those params.
If now they read from ```fillDescriptions``` the customize can of course be removed.

Comment (mmusich) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:277]: >  If now they read from `fillDescriptions` the customize can of course be removed.

yes, it should be the case.

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 3]
```diff
@@ -268,6 +268,7 @@ def customizeHLTfor44576(process):
         break
     return process
 
+
```
Comment (mmusich) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py:271]: if you remove also this line, we should loose `hlt` signature.

**Comment by mmusich** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py (hunk 3 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: OK, I was wrong, becasue

> RecoMuon/L3TrackFinder (hlt, reconstruction) 

sorry for that.

**Comment by slava77** [comment]
@cms-sw/hlt-l2 @cms-sw/reconstruction-l2 
are there any pending comments still?
Please check/clarify.
Thank you.

**Comment by jfernan2** [comment]
+1

**Comment by mmusich** [comment]
HLT was planning to sign after reco signs off (only obligatory config changes in HLT related packages)

**Comment by slava77** [comment]
@rappoccio @antoniovilela 
the PR is fully signed for two days now.
What is the reason to further delay merging?
Please clarify.
Thank you.

**Comment by rappoccio** [comment]
+1

**Pull request merged**